multi: respect and update min relay fee

ellemouton opened the pull request:
Fixes #5390 

This PR ensures that an updated min relay fee is taken into account when updating the commitment fees in a channel. Currently this fee is not considered when updating commitment fees which could result in  commitment txs not being accepted into mempools. This pr does 2 things: 1) respects min relay fee when deciding if commitment fees should be updated and 2) introduces a new minRelayFeeManager which ensures that the relay fee is updated occasionally by querying the chain backend. Currently the minRelayFee is fetched once from the backend at start up and then is not queried again. 

Fixes: https://github.com/lightningnetwork/lnd/issues/4912

Crypt-iQ commented on htlcswitch/link.go:
```diff
@@ -1092,14 +1092,19 @@ func (l *channelLink) htlcManager() {
+			)
 			newCommitFee := chainfee.SatPerKWeight(
-				math.Min(float64(netFee), float64(maxFee)),
+				math.Min(float64(minCommitFee), float64(maxFee)),
```
I think if min relay fee is above the anchor fee, we should choose min relay fee.  Right now anchor max is 10 sats/vbyte

wpaulino commented on lnwallet/chainfee/estimator.go:
```diff
@@ -751,3 +751,64 @@ func (w *WebAPIEstimator) feeUpdateManager() {
+
+	newMinFee, err := m.fetchNewFee()
+	if err != nil {
+		log.Errorf("unable to fetch updated relay fee chain " +
```
Nit: Unable*

wpaulino commented on lnwallet/chainfee/estimator.go:
```diff
@@ -311,11 +311,19 @@ func NewBitcoindEstimator(rpcConfig rpcclient.ConnConfig, feeMode string,
+
+	relayFee := &minRelayFeeManager{
+		fetchNewFee:       b.fetchMinRelayFee,
+		minUpdateInterval: time.Minute * 10,
```
Nit: extract `time.Minute * 10` into a constant so it can be reused in the `BtcdEstimator`.

wpaulino reviewed (commented):
I wonder if comparing against just the minimum relay fee is enough, as we also need to make sure the commitment transaction can enter our mempool first so that it can be relayed to others. Some users out there run with the bitcoind `maxmempool` config option, which restricts the size of their node's mempool, keeping the highest transactions around. In that case, we'd want to be above the minimum mempool fee. cc @Roasbeef

ellemouton commented on htlcswitch/link.go:
```diff
@@ -1101,7 +1110,18 @@ func (l *channelLink) htlcManager() {
+			// If the new commitment fee is below the minimum
+			// relay fee then we bump it to be equal to the minimum
+			// relay fee.
+			minRelayFee := l.cfg.FeeEstimator.RelayFeePerKW()
+			if newCommitFee < minRelayFee {
+				newCommitFee = minRelayFee
+			}
+
+			if !shouldAdjustCommitFee(
+				newCommitFee, commitFee, minRelayFee,
+			) {
```
My only worry here is the extreme case where the minRelayFee is larger than the maximum fee we are able to pay. What do we want to do in that case?

ellemouton replied:
Ok i have updated it to use max fee in this case

ellemouton reviewed (commented):
Thanks for the review guys. Updated 👍  left 1 question

> Some users out there run with the bitcoind `maxmempool` config option

hmmmm yeah good point. Looked into this now and looks like bitcoind has a `getmempoolinfo` rpc that has:
```
 "mempoolminfee" : n,      (numeric) Minimum fee rate in BTC/kB for tx to be accepted. Is the maximum of minrelaytxfee and minimum mempool fee
 ```
 so we could just use this value for bitcoind. is this not something we need to worry about for btcd? I dont see a `maxmempool` config option for btcd

Roasbeef commented:
>  is this not something we need to worry about for btcd? I dont see a maxmempool config option for btcd

Yeah `btcd` doesn't implement the floating floor in the mempool, so it's only something we'll need to worry about for `bitcoind`.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -6916,16 +6913,52 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
+	// If the channel has anchor outputs then we cap the fee at the max
+	// anchor commitment fee rate.
+	if lc.channelState.ChanType.HasAnchors() &&
+		maxFee > maxAnchorCommitFeeRate {
```
On another pass the shift makes more sense, as since we may shift our fee rate up (above the max anchor fee), we need to be able to respect the current configured max fee allocation. However this begs the question of what we should do if the max fee allocation is actually _below_ the min relay fee rate. I think this would be pretty rare in practice, unless someone "misconfigured" their channel fee related settings.

Roasbeef replied:
I think `maxFee` here should be `commitFee` as well. Generally I think this function can be simplified somewhat.

Crypt-iQ replied:
I don't think this would be rare, maybe unexpected but it could happen depending on the local balance, dust limit satoshis, and the min relay fee.

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager.go:
```diff
@@ -0,0 +1,67 @@
+// initMinRelayFee is used to set the minFeePerKW of the minRelayFeeManager
+// and it requires that the call the the backend succeeds. This is just
+// required when the minRelayFeeManager is being initialized.
+func (m *minRelayFeeManager) initMinRelayFee() error {
```
Why not have this just be part of the constructor? So `newMinRelayFeeManager`.

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager.go:
```diff
@@ -0,0 +1,67 @@
+
+	newMinFee, err := m.fetchNewFee()
+	if err != nil {
+		log.Errorf("Unable to fetch updated relay fee chain " +
```
Should log the error here.

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager.go:
```diff
@@ -0,0 +1,67 @@
+	return nil
+}
+
+// getFee returns the stored minFeePerKW if it has been updated recently or
```
`fetchMinFee`?

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager.go:
```diff
@@ -0,0 +1,67 @@
+	minFeePerKW       SatPerKWeight
+	lastUpdatedTime   time.Time
+	minUpdateInterval time.Duration
+	fetchNewFee       func() (SatPerKWeight, error)
```
"new fee" here, being the current min relay fee?

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager_test.go:
```diff
@@ -0,0 +1,54 @@
+	require.Equal(t, minRelayFee, SatPerKWeight(1000))
+	require.Equal(t, 1, chainBackend.callCount)
+
+	time.Sleep(200 * time.Millisecond)
```
One way to get rid of the sleep here would just be updating the last update time in place.

Roasbeef commented on lnwallet/chainfee/minrelayfeemanager_test.go:
```diff
@@ -0,0 +1,54 @@
+// TestMinRelayFeeManager tests that the minRelayFeeManager returns an up to
+// date min relay fee by querying the chain backend and that it returns a cached
+// fee if the chain backend was recently queried.
+func TestMinRelayFeeManager(t *testing.T) {
```
Can add `t.Parallel()` here since it doesn't touch any global state.

Roasbeef commented on lnwallet/chainfee/estimator.go:
```diff
@@ -324,43 +332,35 @@ func NewBitcoindEstimator(rpcConfig rpcclient.ConnConfig, feeMode string,
+	return b.minFeePerKW.initMinRelayFee()
+}
+
+func (b *BitcoindEstimator) fetchMinRelayFee() (SatPerKWeight, error) {
```
Missing a `godoc` comment.

Roasbeef commented on lnwallet/chainfee/estimator.go:
```diff
@@ -324,43 +332,35 @@ func NewBitcoindEstimator(rpcConfig rpcclient.ConnConfig, feeMode string,
 	relayFee, err := btcutil.NewAmount(info.RelayFee)
 	if err != nil {
-		return err
+		return 0, err
```
So I think we actually want to read `MempoolMinFee` here: https://chainquery.com/bitcoin-cli/getmempoolinfo

ellemouton reviewed (commented):
Thanks for the review @Roasbeef 🚀  have updated nits, changed bitcoind to use `mempoolminfee` instead of minRelayFee and also slightly simplified the `IdealCommitFee` function. 

I [was also wondering](https://github.com/lightningnetwork/lnd/pull/5483#discussion_r684040949) about the extreme case when the min mempool fee is actually above the max possible fee allocation. In this pr, if this case happens then the max fee alloc is used since i guess that is the best we can do? seems better to update to the best possible fee than to leave the fee as is?

ellemouton commented on lnwallet/chainfee/minrelayfeemanager.go:
```diff
@@ -0,0 +1,72 @@
+// required by a transaction to be accepted to the mempool.
+// The minRelayFeeManager ensures that the backend used to fetch the fee is not
+// queried too regularly.
+type minRelayFeeManager struct {
```
Would `minFeeManager` be a more appropriate name since we now use `minmempoolfee` for bitcoind but `minrelayfee` for btcd?

Crypt-iQ commented on lnwallet/channel_test.go:
```diff
@@ -8038,18 +8074,33 @@ func TestChannelMaxFeeRate(t *testing.T) {
+	// use the min relay fee.
+	assertIdealFeeRate(anchorChannel, 700000000, 400000, 300000, 0.001, 400000)
+
+	// If the min relay fee is above the maximum fee, use the maximum fee.
```
I think we should override the maximum fee if it's under the min relay fee so the commitment tx has a chance of getting confirmed

Crypt-iQ reviewed (commented):
Just reviewed the first commit -- my opinion is the link fee allocation should be overridden in case it's under min relay fee.  The link fee allocation is a little odd as-is, because if our local balance is 4000sats, we have a 2500sats commitfee, 330sats dust limit satoshis, we will not update the fee to above 4000+2500/2 = 3250 sats with the default max allocation.  If the min relay fee is higher and we need to burn our balance, we should do so as we may have HTLC's to time-out.

Crypt-iQ commented:
An amendment to my original comment: I think if we *can't* actually send min relay fee by burning our local balance, don't bother sending the `update_fee`, and maybe log a warning for the user.

Also, was it the case that the min relay fee was only fetched *once* by the backend (btcd/bitcoind)?

ellemouton commented:
Thanks for taking a look @Crypt-iQ :)

so yeah the question seems to be, if the maxfee (with the set max fee allocation) is less than the relay fee, can we ignore the set max fee alloc and just make it 1 so that we can use the max possible amount? and then the second question is, if this final max fee is still lower than the relay fee, what to do?

>don't bother sending the update_fee, and maybe log a warning for the user.

 surely it might still be worth sending `update_fee` so that we are giving the commitment tx the best change of getting confirmed sooner?

> Also, was it the case that the min relay fee was only fetched once by the backend (btcd/bitcoind)?

yebo yes. currently this fee is fetched once on start up.

ellemouton commented on lnwallet/channel.go:
```diff
@@ -6916,16 +6913,54 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
+	// In the case were the commitment fee is more than the maximum fee
+	// we are able to pay then we set the new commitment fee equal to this
+	// max fee.
+	if commitFee > maxFee {
```
if this check is switched with the min relay fee check then it wont do anything since the switch above it caps commit fee at max fee. But yeah, I will change this up a bit depending on the response to my questions [here](https://github.com/lightningnetwork/lnd/pull/5483#issuecomment-898362068)

Crypt-iQ replied:
Maybe my comment wasn't clear.

Currently:
- switch statement caps at maxfee
- happens to be < minRelayFee, set to minRelayFee
- now > maxFee, set to maxFee (which is < minRelayFee)

My suggestion:
- switch statement caps at maxfee
- won't be > maxFee, do nothing
- may be < minRelayFee, set to minRelayFee

Unless I'm missing something?

ellemouton replied:
see latest update :)

Crypt-iQ commented:
> surely it might still be worth sending update_fee so that we are giving the commitment tx the best likely hood of being relayed if the relay fee does eventually drop?

I agree

ellemouton commented:
Cool i have updated the PR so that:
- if commit fee is < min relay fee && our ideal max fee (ie with maxFeeAlloc) is > relay fee, then use relay fee
- else if absolute max fee (setting maxFeeAlloc=1) is > relay fee, then use relay fee and log a warning to say we are ignoring the maxFeeAlloc
- else if absolute max fee is < relay fee, then do best effort and update fee to be that absolute max fee and log error

ellemouton commented on lnwallet/channel.go:
```diff
@@ -6916,16 +6913,75 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
+	lc.log.Error("The commitment fee rate is below the current " +
+		"minimum relay fee rate.")
+
+	return absoluteMaxFee
```
Ok i think this should be addressed now in the first 2 commits. They ensure that MaxFeeRate takes the reserve into account. I also think I found a bug in `validateFeeRate` that should be fixed now.

Roasbeef replied:
I think this possibility of "use all the users fees for the commitment transaction", is kinda scary, but then again it shouldn't really ever happen since it's unlikely that the min relay fee ever gets this high. 

One thing we can do to soften the blow here is to attempt to use the smallest max fee possible here. So instead of start at 100%, we increase in an iterative manner from the current config value, up to the max fee if that gets us above the fee rate.

ellemouton replied:
> One thing we can do to soften the blow here is to attempt to use the smallest max fee possible here

Is this referring to just line 6982 or also to check at 6967? If the latter, then does that mean that even if we can meet the min relay fee initially if we were to use maxAlloc=1 we shouldn't? If the former, then do you mean something like we just take the last used commitment fee and bump by 10% or something until we reach absolute max?

Roasbeef replied:
> then does that mean that even if we can meet the min relay fee initially if we were to use maxAlloc=1 we shouldn't?

I guess I'm referring more so between finding the sweet spot between the minrelay fee and the absolute maximum fee. As a basic example, let's say the max fee is 20 and the min relay fee is 5 (larger gap just to illustrate the point). With this logic we'd just to 20 rather than picking say 6.

ellemouton replied:
hmm in the example you give, this logic will return 5. The absolute max is only ever returned iff the absolute max is strictly smaller than the min relay fee

ellemouton commented on lnwallet/channel_test.go:
```diff
@@ -8014,6 +8014,10 @@ func TestChannelMaxFeeRate(t *testing.T) {
 				maxAlloc, maxFeeRate)
 		}
+
+		if err := c.validateFeeRate(maxFeeRate); err != nil {
+			t.Fatalf("fee rate validation failed: %v", err)
+		}
```
note that before this commit, adding this validation check here would fail the test

ellemouton replied:
ok i have tried this. See `propertyTest` below.

ellemouton commented:
Thanks @Crypt-iQ ! I have updated. The `MaxFeeRate` function now takes our local reserve amount into account. I also added a commit that fixes the `validateFeeRate` function. Before, this function would fail the MaxFeeRate test if called there since it seems to incorrectly calculate the base balance.

Crypt-iQ reviewed (commented):
Think this is almost there, just a comment about `validateFeeRate`

ellemouton commented on lnwallet/channel.go:
```diff
@@ -6715,12 +6715,19 @@ func (lc *LightningChannel) validateFeeRate(feePerKw chainfee.SatPerKWeight) err
+	_, txWeight := lc.availableBalance()
+
+	localBalance := lnwire.NewMSatFromSatoshis(
+		lc.channelState.LocalCommitment.LocalBalance.ToSatoshis(),
```
ok sure, but then i think that we should also then use `availableBalance` in the `MaxFeeRate` function so that it also takes the pending htlc's into account. Otherwise `validateFeeRate` can not reliably be used to test the edge cases returned by `MaxFeeRate`.  Furthermore, I think that if we do use the `availableBalance` function in both MaxFeeRate and validateFeeRate, then we can remove the reserve balance checks in both since `availableBalance` already handles this for us

ellemouton commented on lnwallet/channel.go:
```diff
@@ -6738,17 +6741,6 @@ func (lc *LightningChannel) validateFeeRate(feePerKw chainfee.SatPerKWeight) err
 			newFee, baseBalance)
 	}
 
-	// If this new balance is below our reserve, then we can't accommodate
-	// the fee change, so we'll reject it.
-	balanceAfterFee := baseBalance - newFee
-	if balanceAfterFee.ToSatoshis() < lc.channelState.LocalChanCfg.ChanReserve {
-		return fmt.Errorf("cannot apply fee_update=%v sat/kw, "+
-			"new balance=%v would dip below channel reserve=%v",
-			int64(feePerKw),
-			balanceAfterFee.ToSatoshis(),
-			lc.channelState.LocalChanCfg.ChanReserve)
-	}
-
```
I have removed this since i dont think it is possible for the `balanceAfterFee` to be less than our local reserve amount. The `availableBalance` already takes into account our reserve amount and so we would actually need to add the reserve amount to balanceAfterFee for this check to mean anything. And if we did so, then then only way this check would fail is if `baseBalance - newFee` was negative, in which case, the previous check would have errored.

Roasbeef replied:
IIRC we added this to guard against an actual issue we encountered in the wild. Would prefer to leave it in place based on that pretext: https://github.com/lightningnetwork/lnd/commit/3e90000e2af2cd8212762e4de6419a43977d7984 

Also a quick log search on my mainnet node shows that hits is indeed being hit in the wild: 
```
$ ack 'dip below' .  | wc
 68    1564   14675
```

ellemouton replied:
This check was added before the change in [5e89d5b](https://github.com/lightningnetwork/lnd/commit/5e89d5b) which adds [this](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L6656). So the availableBalance returned now already has the reserve amount subtracted from it which was not the case when this check was originally added. I think the reason why the log is currently still being hit is because `balanceAfterFee` already has the reserve amount subtracted from it so it is very possible that this balance is below reserve, but that is ok cause reserve was already subtracted. So i dont think this check makes sense unless `balanceAfterFee` is changed to `baseBalance - newFee + chanReserve`. which will always be >= chanReserve.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -6912,16 +6909,77 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
+	// relay fee rate. The commitment tx will not be able to propagate.
+	// To give the transaction the best chance, we use the absolute
+	// maximum fee we have available and we log an error.
+	lc.log.Error("The commitment fee rate is below the current " +
```
Log the current fee, the max, and the relay fee here?

Roasbeef commented on lnwallet/channel_test.go:
```diff
@@ -8028,12 +8064,21 @@ func TestChannelMaxFeeRate(t *testing.T) {
+		aliceChannel, 500000000, 600000000, 0, 1.0, 600000000,
+	)
+
+	// Test ideal fee rates for an anchor channel
```
Could make this into a sub-test: https://pkg.go.dev/testing#hdr-Subtests_and_Sub_benchmarks

Crypt-iQ replied:
I don't know that we can combine the logic here with `assertIdealFeeRate` as there are differing values which is one of the main advantages of subtests.  I think `CreateTestChannels` could be combined, but not sure it has a true benefit.

Roasbeef commented on lnwallet/chainfee/estimator.go:
```diff
@@ -406,7 +410,7 @@ func (b *BitcoindEstimator) EstimateFeePerKW(
 // NOTE: This method is part of the Estimator interface.
 func (b *BitcoindEstimator) RelayFeePerKW() SatPerKWeight {
-	return b.minFeePerKW
+	return b.minFeeManager.fetchMinFee()
```
@Crypt-iQ re your comment the other day, your statement is _true_ after this commit

Crypt-iQ replied:
👍

Roasbeef reviewed (approved):
LGTM 🍬

Great work on this! As usual, any sort of fee handling off-chain is rife with countless edge cases. Final few comments re some unlikely behavior and some additional logging as well as an iterative approach we can use to soften the blow in the worst case.

ellemouton commented:
Thanks for the review @Roasbeef 🚀 I just left one question to ask for some clarification on your comment.

Crypt-iQ reviewed (changes requested):
just some nits

Crypt-iQ commented on lnwallet/chainfee/estimator.go:
```diff
@@ -164,34 +164,36 @@ func (b *BtcdEstimator) Start() error {
+// fetchMinRelayFee fetches and returns the minimum relay fee in sat/kb from
+// the btcd backend.
+func (b *BtcdEstimator) fetchMinRelayFee() (SatPerKWeight, error) {
+	info, err := b.btcdConn.GetInfo()
```
@Roasbeef so this doesn't float according to mempool conditions? If that's true, then perhaps this doesn't need to be refreshed as it's effectively static?

Roasbeef replied:
Correct, but no harm leaving it in IMO, since it makes the logic a bit more uniform.

ellemouton commented on lnwallet/channel.go:
```diff
@@ -6905,16 +6906,19 @@ func (lc *LightningChannel) MaxFeeRate(maxAllocation float64,
+	// balance.
+	availableBalance, weight := lc.availableBalance()
+
+	oldFee := lc.localCommitChain.tip().feePerKw.FeeForWeight(weight)
```
hmmm yeah so a bit of a race. But it was also not 100% accurate before right?

Crypt-iQ replied:
Yeah wasn't accurate before either

Crypt-iQ commented on lnwallet/channel_test.go:
```diff
@@ -8025,60 +8031,165 @@ func TestChannelMaxFeeRate(t *testing.T) {
+	// inputs fed to MaxFeeRate.
+	propertyTest := func(c *LightningChannel) func(ma maxAlloc) bool {
+		return func(ma maxAlloc) bool {
+			maxFeeRate := c.MaxFeeRate(float64(ma))
```
Should this use `IdealCommitFeeRate` instead? O/w it's the same as the other property test

ellemouton replied:
Indeed! good catch

Crypt-iQ commented on lnwallet/channel_test.go:
```diff
@@ -8005,12 +8005,18 @@ func TestForceCloseBorkedState(t *testing.T) {
 func TestChannelMaxFeeRate(t *testing.T) {
```
I think this should also be tested for anchor channels

Crypt-iQ commented on lnwallet/channel_test.go:
```diff
@@ -8025,60 +8031,165 @@ func TestChannelMaxFeeRate(t *testing.T) {
+	// Test ideal fee rates for an anchor channel
+	t.Run("anchor-channel", func(t *testing.T) {
+		anchorChannel, _, cleanUp, err := CreateTestChannels(
+			channeldb.SingleFunderTweaklessBit |
```
Could add the `ZeroFeeHtlcTx` bit, all anchor chans lnd makes are now zero-fee

Crypt-iQ reviewed (commented):
This is mostly there 🦖 , just some test-related comments. Will run a for-loop bash script to see if any flakes occur with the property-based testing and report back

Crypt-iQ commented on lnwallet/chainfee/minfeemanager.go:
```diff
@@ -0,0 +1,74 @@
+// given fetchMinFee function to set the minFeePerKW of the minFeeManager.
+// This function requires the fetchMinFee function to succeed.
+func newMinFeeManager(minUpdateInterval time.Duration,
+	fetchMinFee func() (SatPerKWeight, error)) (*minFeeManager, error) {
```
nit: could change `func() (SatPerKWeight, error)` -> `fetchFee`

ellemouton commented:
updated as per comments @Crypt-iQ 🚀
